### ISSUE TITLE
Precision and Recall always be 0.

### ISSUE THREAD

**Issue by mimi37** [ultralytics/yolov5#9038]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training

### Bug

I run the template from the website“python train.py --data coco128.yaml --weights yolov5s.pt --img 640”, however the precision and recall has always been 0. I am sure I pull the newest version which means the train.py has been modified.
![截屏2022-08-19 23 34 41](https://user-images.githubusercontent.com/41326307/185654967-5533b882-0502-4ac9-9fc2-25b619165e0a.png)

### Environment

exactly as same as the requirements.txt

### Minimal Reproducible Example

no

### Additional

 no

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by li221199** [comment]
I have the same problem, the precision and recall has always been 0
what is the cause and how to solve it?
![image](https://user-images.githubusercontent.com/38066925/185655963-088ee4df-4dcd-431e-8b9a-0c38321fa7b9.png)

**Comment by elvecinodelquinto** [comment]
Same problem here. Seems it is not calculating anything for the validation subset.

**Comment by triple-Mu** [comment]
It seems that bn has someting wrong when training model.
![image](https://user-images.githubusercontent.com/92794867/185728239-4bf95cda-5387-41f3-81ec-40619b5a170e.png)

**Comment by robotwhispering** [comment]
I've been chasing this problem all day long. I did a git pull today because it said i was behind, and then I've had P, R, and mAP show either 0 or NaN, depending on the chart, no matter what I do. I've uninstalled my virtual environment, reinstalled, tried different versions of pytorch, etc.
![Screenshot from 2022-08-20 02-38-17](https://user-images.githubusercontent.com/60989737/185734700-7176555a-f9d8-4920-b27e-baecc86223d7.png)

**Comment by Th1nhNg0** [comment]
I got the same problem. I have train for 2 week, everything ok until yesterday

**Comment by triple-Mu** [comment]
> I got the same problem. I have train for 2 week, everything ok until yesterday

Now I rollback to 6.1.
It looks like more stable.

**Comment by glenn-jocher** [comment]
@Th1nhNg0 @li221199 @mimi37 @robotwhispering @triple-Mu good news 😃! Your original issue may now be fixed ✅ in PR #9037. This PR resolves a zero-mAP bug.

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by glenn-jocher** [comment]
@Th1nhNg0 @triple-Mu @robotwhispering @mimi37 @li221199 @elvecinodelquinto good news 😃! Your original issue may now be fixed ✅ in PR #9068 by @0zppd. @pourmand1376 tracked down the problem using `git bisect` and running each training 10x since the bug was not reproducible on every training, but showed up in maybe 1/3 of the trainings. 

The original issue was that I had replaced torch.zeros() with torch.empty() on some ops like warmup and profiling to try to get some slight speed improvements, and once op in particular ran a torch.empty() tensor through the model when it was in .train() mode, leading the batchnorm layers to add those values to the tracked statistics. Since torch.empty is not initialized it can take on extremely high or low values, leading some batchnorm layers to randomly output NaN. 

The PR has been extensively tested on 10x Colab trainings and all 10 came back good now:
<img width="1211" alt="Screenshot 2022-08-21 at 15 40 00" src="https://user-images.githubusercontent.com/26833433/185793959-986ad082-c181-49b8-a564-b96270ebc409.png">

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
